src/bookshelf/src/bksw_sec.erl

```diff
@@ -20,128 +21,156 @@
-                                       RequestId, StringToSign, Req0, Context)
+            encode_sign_error_response(AWSAccessKeyId, IncomingSignature, reqid(Auth),
+                                       CalculatedSig, req(Auth), Context)
     end.
 
--spec expire_diff(undefined | binary()) -> integer().
-expire_diff(undefined) -> 1;
-expire_diff(Expires) ->
-    Now = calendar:datetime_to_gregorian_seconds(erlang:universaltime()),
-    bksw_util:to_integer(Expires) - (Now - ?SECONDS_AT_EPOCH).
+%%===================================================================
+% local functions, helpers, etc.
+%%===================================================================
+
+% common setup and init
+-spec auth_init(any(), tuple(), string()) -> map().
+auth_init(Req0, Context, SignedHeaders) ->
+    AccessKey            =  bksw_conf:access_key_id(Context),
+    {RequestId, Req1}    =  bksw_req:with_amz_request_id(Req0),
+    Config               =  mini_s3:new(AccessKey, bksw_conf:secret_access_key(Context), host(Req1), path),
+    #{accesskey          => AccessKey,
+      config             => Config,
+      alt_signed_headers => [case {K, V} of {"host", _} -> {"host", get_host_toggleport(host(Req1), Config)}; _ -> {K, V} end || {K, V} <- SignedHeaders],
+      method             => list_to_atom(string:to_lower(erlang:atom_to_list(wrq:method(Req1)))),
+      path               => wrq:path(Req1),
+      req                => Req1,
+      reqid              => RequestId}.
+
+% due to a security vulnerability described by Mark Anderson, we should compare signatures
+% in constant time, and not 'early out' on the first mismatched character.  this means we
+% are purposefully using a 'deoptimized' string compare function.
+-spec const_time_compare(string() | binary() | atom(), string() | binary() | atom(), boolean()) -> boolean().
+const_time_compare(S1, S2, IsEqual) when is_binary(S1) ->
+    const_time_compare(binary_to_list(S1), S2, IsEqual);
+const_time_compare(S1, S2, IsEqual) when is_binary(S2) ->
+    const_time_compare(S1, binary_to_list(S2), IsEqual);
+const_time_compare(S1, S2, IsEqual) when is_list(S1), is_list(S2) ->
+    ctcomp(S1, S2, IsEqual);
+const_time_compare(_, _, _) ->
+    false.
 
-%% get_bucket([Bucket, _, _]) ->
-%%     Bucket.
+-spec ctcomp(string(), string(), boolean()) -> boolean().
+ctcomp([    ], [    ], IsEqual) -> IsEqual;
+ctcomp([    ], [_|S2],       _) -> ctcomp([], S2, false  );
+ctcomp([_|S1], [    ],       _) -> ctcomp(S1, [], false  );
+ctcomp([X|S1], [X|S2], IsEqual) -> ctcomp(S1, S2, IsEqual);
+ctcomp([_|S1], [_|S2],       _) -> ctcomp(S1, S2, false  ).
+
+encode_access_denied_error_response(RequestId, Req0, Context) ->
+    Req1 = bksw_req:with_amz_id_2(Req0),
+    Body = bksw_xml:access_denied_error(RequestId),
+    Req2 = wrq:set_resp_body(Body, Req1),
+    {{halt, 403}, Req2, Context}.
 
 encode_sign_error_response(AccessKeyId, Signature,
                            RequestId, StringToSign, Req0,
```

Comment:
Should this variable name change, if this isn't a StringToSign?

Reply:
Signature => IncomingSignature
StringToSign => RecreatedSignature

Done.